### PR TITLE
 curvefs/client: deal with some remaining issues

### LINKED ISSUE #737: curvefs/client: deal with some remaining issues
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**
1.flush finds that the file does not exist, then clean up all caches
2. The logic of UpdateWrteCacheMap here has a reverse call, and it leads to
chunkCacheManager_->rwLockWrite_ becomes public and needs to be optimized
3. Truncate becomes smaller, and the corresponding cache needs to be released
4. Flush fails, and there is a scene where a new write cache is added. The current version of the code is already supported, just add unit tests

**Describe the solution you'd like (描述你期望的解决方法)**

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by baijiaruo**
### What problem does this PR solve?
1. If the flushing file is removed, then clean up all caches
2. The logic of UpdateWriteCacheMap here has a reverse call, and it leads to
chunkCacheManager_->rwLockWrite_ becomes public and needs to be optimized
3. Truncate becomes smaller, and the corresponding cache needs to be released
Issue Number: close #737  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 1]
```diff
@@ -1239,7 +1321,7 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force) {
     curve::common::LockGuard lg(flushMtx_);
     CURVEFS_ERROR ret;
     {
-        WriteLockGuard writeLockGuard(rwLockWrite_);
+        WriteLockGuard writeLockGuard(rwLockChunk_);
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1324]: why change to another lock?

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 2]
```diff
@@ -1594,7 +1667,6 @@ void DataCache::Write(uint64_t chunkPos, uint64_t len, const char *data,
                 */
                 if (chunkPos + len <
                     (*iter)->GetChunkPos() + (*iter)->GetLen()) {
-                    chunkCacheManager_->rwLockWrite_.WRLock();
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1597]: you just remove this lock, and `chunkCacheManager_->UpdateWriteCacheMap` won't hold lock when modify `dataWCacheMap_`, is this correct?

**Comment by baijiaruo** [comment]
recheck

**Comment by baijiaruo** [comment]
recheck

**Comment by YunhuiChen** [comment]
recheck

**Comment by YunhuiChen** [comment]
recheck

**Comment by baijiaruo** [comment]
recheck

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 3]
```diff
@@ -827,13 +838,35 @@ void FileCacheManager::ReleaseCache() {
...
     return;
 }
 
+void FileCacheManager::TruncateCache(uint64_t offset, uint64_t fileSize) {
+    uint64_t chunkSize = s3ClientAdaptor_->GetChunkSize();
+    uint64_t chunkIndex = offset / chunkSize;
+    uint64_t chunkPos = offset % chunkSize;
+    int chunkLen = 0;
+    while (fileSize > 0) {
+        if (chunkPos + fileSize > chunkSize) {
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:854]: Add some code comments

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 4]
```diff
@@ -136,11 +136,22 @@ CURVEFS_ERROR FsCacheManager::FsSync(bool force) {
...
+            if (iter1 != fileCacheManagerMap_.end()) {
+                VLOG(9) << "Release FileCacheManager, inode id: "
+                        << iter1->second->GetInodeId();
+                fileCacheManagerMap_.erase(iter1);
+            }
+        } else {
+            LOG(ERROR) << "fs fssync error, ret: " << ret;
             return ret;
         }
         /* TODO(huyao): need delete source, now it has race condition
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:157]: remove the deprecated code

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 5]
```diff
@@ -827,13 +838,35 @@ void FileCacheManager::ReleaseCache() {
...
+    while (fileSize > 0) {
+        if (chunkPos + fileSize > chunkSize) {
+            chunkLen = chunkSize - chunkPos;
+        } else {
+            chunkLen = fileSize;
+        }
+        ChunkCacheManagerPtr chunkCacheManager =
+            FindOrCreateChunkCacheManager(chunkIndex);
+        chunkCacheManager->TruncateCache(chunkPos);
+        fileSize -= chunkLen;
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:862]: Why TruncateCache One by One, the  hole chunk of  (offset / chunkSize+1,  fileSize/chunkSize) should be throw

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 6]
```diff
@@ -157,60 +157,65 @@ int S3ClientAdaptorImpl::Read(uint64_t inodeId, uint64_t offset,
 CURVEFS_ERROR S3ClientAdaptorImpl::Truncate(Inode* inode, uint64_t size) {
     uint64_t fileSize = inode->length();
 
-    if (size <= fileSize) {
-        VLOG(6) << "Truncate size:" << size << " less or equal fileSize"
+    if (size < fileSize) {
+        VLOG(6) << "Truncate size:" << size << " less or equal fileSize:"
```
[the remark this answers is not in the document]
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_adaptor.cpp:161]: fixed

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 7]
```diff
@@ -827,13 +838,35 @@ void FileCacheManager::ReleaseCache() {
     auto iter = chunkCacheMap_.begin();
 
     for (; iter != chunkCacheMap_.end(); iter++) {
-        iter->second->ReleaseCache(s3ClientAdaptor_);
+        iter->second->ReleaseCache();
     }
```
[the remark this answers is not in the document]
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:842]: fixed

**Comment by baijiaruo** [comment]
recheck

**Comment by baijiaruo** [comment]
recheck

**Comment by YunhuiChen** [comment]
recheck

**Review by xu-chaojie — APPROVED** [review]

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**